File: app/src/main/java/net/myacxy/squinch/helpers/tracking/builder/ExceptionLogBuilder.java
package net.myacxy.squinch.helpers.tracking.builder;

import android.support.annotation.NonNull;

import net.myacxy.squinch.helpers.tracking.ITracker;
import net.myacxy.squinch.helpers.tracking.TrackingHelper;

public class ExceptionLogBuilder extends BaseLogBuilder<ExceptionLogBuilder> {

    private Throwable throwable;

    public ExceptionLogBuilder(TrackingHelper trackingHelper) {
        super(trackingHelper);
    }

    public ExceptionLogBuilder withThrowable(Throwable throwable) {
        this.throwable = throwable;
        return this;
    }

    @Override
    public TrackingHelper post() {
        for (ITracker tracker : trackers) {
            tracker.exception(throwable);
        }
        return trackingHelper;
    }

    @NonNull
    @Override
    protected ExceptionLogBuilder self() {
        return this;
    }
}

File: app/src/main/java/net/myacxy/squinch/helpers/tracking/builder/ExceptionLogBuilder.kt
package net.myacxy.squinch.helpers.tracking.builder

import net.myacxy.squinch.helpers.tracking.TrackingHelper

class ExceptionLogBuilder(trackingHelper: TrackingHelper) : BaseLogBuilder<ExceptionLogBuilder>(trackingHelper) {

    private var throwable: Throwable? = null

    fun withThrowable(throwable: Throwable?): ExceptionLogBuilder {
        this.throwable = throwable
        return self()
    }

    override fun post(): TrackingHelper {
        trackers.forEach { it.exception(throwable) }
        return trackingHelper
    }

    override fun self(): ExceptionLogBuilder {
        return this
    }
}
